Firebase.h

```diff
@@ -25,48 +25,123 @@
 
-// FirebaseError represents a Firebase API error with a code and a
-// message.
+class FirebaseGet;
+class FirebasePush;
+class FirebaseRemove;
+class FirebaseStream;
+
+// Primary client to the Firebase backend.
+class Firebase {
+ public:
+  Firebase(const String& host);
+  Firebase& auth(const String& auth);
+
+  // Fetch result at "path".
+  FirebaseGet get(const String& path);
+
+  // Add new value to list at "path", will return key for the new item.
+  FirebasePush push(const String& path, const String& value);
+
+  // Deletes value at "path" from firebase.
+  FirebaseRemove remove(const String& path);
+
+  // Starts a stream of events that affect object at "path".
+  FirebaseStream stream(const String& path);
+
+ private:
+  HTTPClient http_;
+  String host_;
+  String auth_;
+};
+
 class FirebaseError {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  FirebaseError() {}
+  FirebaseError(int code, const String& message) : code_(code), message_(message) {
+  }  
+  operator bool() const { return code_ != 0; }
+  int code() const { return code_; }
+  const String& message() const { return message_; }
+ private:  
+  int code_ = 0;
+  String message_ = "";
+};
+
+class FirebaseCall {
+ public:
+  FirebaseCall(const String& host, const String& auth,
+	       const char* method, const String& path,
+	       const String& data = "",	       
+	       HTTPClient* http = NULL);
+  const FirebaseError& error() const {
+    return error_;
+  }
+  const String& response() {
+    return response_;
+  }
+ protected:
+  HTTPClient http_;
+  FirebaseError error_;
+  String response_;
+};
+
+class FirebaseGet : public FirebaseCall {
+ public:
+  FirebaseGet(const String& host, const String& auth,
+	      const String& path, HTTPClient* http = NULL);
+  
+  const String& json() const {
+    return json_;
   }
+
  private:
-  int _code = 0;
-  String _message = "";
+  String json_;
 };
 
-// Firebase is the connection to firebase.
-class Firebase {
+class FirebasePush : public FirebaseCall {
  public:
-  Firebase(const String& host);
-  Firebase& auth(const String& auth);
-  const FirebaseError& error() const {
-    return _error;
+  FirebasePush(const String& host, const String& auth,
+	       const String& path, const String& value, HTTPClient* http = NULL);
+
+  const String& name() const {
+    return name_;
   }
-  String get(const String& path);
-  String push(const String& path, const String& value);
-  bool connected();
-  Firebase& stream(const String& path);
+
+ private:
+  String name_;
+};
+
+class FirebaseRemove : public FirebaseCall {
+ public:
+  FirebaseRemove(const String& host, const String& auth,
+		 const String& path, HTTPClient* http = NULL);
+};
+
+
+class FirebaseStream : public FirebaseCall {
+ public:
```

Comment:
We should probably allow a lazy construction of the `FirebaseStream` object (and even the other type) through a `begin()` method, that seems to be a popular pattern in the Arduino world, and would solve some of the lifecycle headache.